app/src/main/kotlin/compiler/intermediate/ControlFlow.kt

```diff
@@ -363,20 +405,94 @@ object ControlFlow {
                     last = lastAfterTrueBranch + lastAfterFalseBranch
                     IFTNode.RegisterRead(resultTemporaryRegister)
                 }
-                is Expression.ArrayLength -> TODO()
-                is Expression.ArrayElement -> TODO()
-                is Expression.ArrayAllocation -> TODO()
+                is Expression.ArrayLength -> {
+                    val array = makeCFGForSubtree(astNode.expression)
+                    val resultReg = Register()
+                    cfgBuilder.addNextTree(
+                        IFTNode.RegisterWrite(
+                            resultReg,
+                            IFTNode.MemoryRead(IFTNode.Subtract(array, IFTNode.Const(memoryUnitSize.toLong())))
+                        )
+                    )
+                    cfgBuilder.addNextCFG(arrayMemoryManagement.genRefCountDecrement(array, expressionTypes[Ref(astNode.expression)]!!))
```

Comment:
Where is the corresponding increment?

Reply:
It is done while evaluating `astNode.expression` to obtain `array` in the first line of the function.